pydy/codegen/ode_function_generator.py

```diff
@@ -0,0 +1,705 @@
+                                                         in f(q, u, r, p)])
+
+    def generate_min_mass_matrix_function(self):
+
+        self._define_inputs()
+        outputs = [self.mass_matrix, self.right_hand_side,
+                   self.coordinate_derivatives]
+
+        f = self._lambdify(outputs)
+
+        if self.specifieds is None:
+            self.eval_arrays = lambda q, u, p: tuple([np.squeeze(o) for o in
+                                                      f(q, u, p)])
+        else:
+            self.eval_arrays = lambda q, u, r, p: tuple([np.squeeze(o) for o
+                                                         in f(q, u, r, p)])
+
+
+class TheanoODEFunctionGenerator(ODEFunctionGenerator):
+
+    def __init__(self, *args, **kwargs):
+
+        if theano is None:
+            raise Exception('Theano must be installed to use this class.')
+        else:
+            super(TheanoODEFunctionGenerator, self).__init__(*args, **kwargs)
+
+    def _define_inputs(self):
+
+        if self.specifieds is None:
+            self.inputs = self.coordinates + self.speeds + self.constants
+        else:
+            self.inputs = (self.coordinates + self.speeds + self.specifieds
+                           + self.constants)
+
+    def _theanoize(self, outputs):
+
+        self._define_inputs()
+
+        f = theano_function(self.inputs, outputs, on_unused_input='ignore')
+
+        # Theano will run faster if you trust the input. I'm not sure
+        # what the implications of this are. See:
+        # http://deeplearning.net/software/theano/tutorial/faq.html#faster-small-theano-function
+        # Note that map(np.asarray, np.hstack(args)) is required if
+        # trust_input is True. If it is False, then it will sanitize the
+        # inputs. I'm not sure which one is faster.
+        f.trust_input = True
+
+        return f
+
+    def generate_full_rhs_function(self):
+
+        outputs = [self.right_hand_side]
+
+        f = self._theanoize(outputs)
+
+        def eval_arrays(*args):
+            vals = map(np.asarray, np.hstack(args))
+            return np.squeeze(f(*vals))
+
+        self.eval_arrays = eval_arrays
+
+    def generate_full_mass_matrix_function(self):
+
+        outputs = [self.mass_matrix, self.right_hand_side]
+
+        f = self._theanoize(outputs)
+
+        def eval_arrays(*args):
+            vals = map(np.asarray, np.hstack(args))
+            return tuple([np.squeeze(o) for o in f(*vals)])
+
+        self.eval_arrays = eval_arrays
+
+    def generate_min_mass_matrix_function(self):
+
+        outputs = [self.mass_matrix, self.right_hand_side,
+                   self.coordinate_derivatives]
+
+        f = self._theanoize(outputs)
+
+        def eval_arrays(*args):
+            vals = map(np.asarray, np.hstack(args))
+            return tuple([np.squeeze(o) for o in f(*vals)])
+
+        self.eval_arrays = eval_arrays
+
+
+def generate_ode_function(right_hand_side, coordinates, speeds, constants,
```

Comment:
I assume this is the intended entry point for the user, and all the above classes are just implementation?